edgelet/edgelet-docker/src/image_prune_data.rs

```diff
@@ -0,0 +1,635 @@
+        &self,
+        in_use_image_ids: HashSet<String>,
+    ) -> Result<HashMap<String, Duration>, Error> {
+        let guard = self
+            .inner
+            .lock()
+            .map_err(|e| Error::LockError(e.to_string()))?;
+
+        let settings = guard.settings.clone();
+
+        // Read persistence file into in-mem map. This map now contains
+        // all images deployed to the device (through an IoT Edge deployment).
+        // If persistence file cannot be read we will return a new map so
+        // new persistence file will be created.
+        let iotedge_images_map = match get_images_with_timestamp(guard.image_use_filepath.clone()) {
+            Ok(map) => map,
+            Err(e) => {
+                drop(guard);
+                log::warn!("Could not read image garbage collection data. Image garbage collection will not prune any images. {}", e);
+                return Ok(HashMap::new());
+            }
+        };
+
+        /* ============================== */
+
+        // process maps
+        let (images_to_delete, carry_over) = process_state(
+            iotedge_images_map,
+            in_use_image_ids,
+            settings.image_age_cleanup_threshold(),
+        )?;
+
+        /* ============================== */
+
+        // write previously removed entries back to file
+        if let Err(e) = write_images_with_timestamp(
+            &carry_over,
+            guard.tmp_filepath.clone(),
+            guard.image_use_filepath.clone(),
+        ) {
+            log::warn!("Failed to update image auto pruning persistence file. File will be updated on next scheduled run. {}", e);
+        };
+
+        /* ============================== */
+
+        drop(guard);
+
+        // these are the images we need to prune; file has already been updated
+        Ok(images_to_delete)
+    }
+}
+
+/* ===================================== HELPER METHODS ==================================== */
+
+fn get_images_with_timestamp(
+    image_use_filepath: String,
+) -> Result<HashMap<String, Duration>, Error> {
+    if !std::path::Path::new(&image_use_filepath).exists() {
+        log::info!(
+            "Image garbage collection data file not found; creating file at: {}",
+            image_use_filepath.as_str()
+        );
+        let _file = fs::File::create(image_use_filepath.clone()).map_err(Error::CreateFile)?;
+    }
+
+    let res = fs::read_to_string(image_use_filepath);
+    if let Err(e) = res {
+        let msg = format!("Could not read image persistence data: {}", e);
+        log::error!("{msg}");
+        return Err(Error::FileOperation(msg));
+    }
+
+    let contents = res.expect("Reading image persistence data failed");
+
+    let mut image_map: HashMap<String, Duration> = HashMap::new(); // all image pruning data
+
+    // TL;DR: this dumps pruning data into the image_map, where
+    // Key: Image hash, Value: Timestamp when image was last used (in epoch)
+    let contents = contents
```

Comment:
Why not use serde to store/parse the file?

Reply:
Didn't think of it. Not taking the suggestion for the reasons mentioned here: https://github.com/Azure/iotedge/pull/6576/files#r949688177